### PR TITLE
replace RList by list R in all files where it is used in this directory

### LINKED ISSUE #6607: Replace Rlist by list R
https://github.com/coq/coq/blob/4d8903c06fd9fd2aca793da8bb55448906347a8c/theories/Reals/RList.v#L13

### LINKED ISSUE #11396: Rlist hides standard list constructors cons and nil in 8.11
In 8.11 we have this:
```
Require Import Reals.
Print cons.

Inductive Rlist : Set := nil : Rlist | cons : R -> Rlist -> Rlist
Arguments cons _%R_scope
```
This results in very substantial trouble in many modules which include Reals. A major issue is that many tactics use the standard list constructors. Unlike for Gallina code Coq cannot check at compile time if using such tactics would actually work - usually it doesn't. Many modules using Reals compile fine (possibly after some patching) but tactics don't work. The required patching can be substantial and tactics must be tested.

Since this is such a mess and likely also breaks a lot of user code (likely pretty much everything whic uses Reals) I would consider replacing Rlist with list R. I had a discussion with @ybertot on gitter on the topic and he mentioned that the raison d'être of Rlist is long gone (list was once restricted to Set).

I think we need to decide if we replace Rlist with list R and fix up the standard library in 8.11. I would strongly vote for it. It is likely faster than fixing up all the remaining failing modules in the 8.11 Windows installer (gappa is failing and proper picking for interval is extremely time consuming, some things are still untested).

### PR DESCRIPTION
**Pull request by ybertot**
This is a first step, ideally In should be List.In, Rlength should be
 List.length, etc.

Nameless intros were only transformed into named intros where necessary
for the proof to resist the change of default names.  Only one instance
of generated name was being re-used in a modified tactic.  In this case,
the culprit name was added explicitely in the corresponding intro.

**Kind:** documentation / bug fix / feature / performance / infrastructure.

Fixes #6607
Fixes / closes #11396

- [ ] Added / updated test-suite

- [ ] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Comment by herbelin** [comment]
Hi @ybertot, there seems to be two PRs in parallel to address #11396! The other PR #11398 is proposing a "minimal" fix and you are proposing a more radical fix which seems justified to me.

In both cases, coquelicot has to be changed (unless in #11404, deprecation aliases for `Rlist.nil` and `Rlist.cons` are provided).

I don't know what the 8.11 manager prefer to do for 8.11.

**Comment by Zimmi48** [comment]
The other PR was opened after discussing the various options and the RM saying that the only reasonable option for 8.11 was the less radical one. Longer term, this one makes more sense.

**Comment by ppedrot** [comment]
Up to documentation / changelog, this seems OK, doesn't it?

**Comment by MSoegtropIMC** [comment]
> Up to documentation / changelog, this seems OK, doesn't it?

Looks good to me, and it is definitely a desirable change. The entities not yet in main CI (interval, gappa) don't use RList directly afaik.

**Comment by ppedrot** [comment]
Given the lack of reaction from @coq/reals-library-maintainers, I am assigning.

**Comment by herbelin** [comment]
> Given the lack of reaction from @coq/reals-library-maintainers, I am assigning.

There were reactions from @coq/reals-library-maintainers but a misunderstanding on my side on the PR being ready (e.g. `Rlength` and `In` are apparently now done). Thanks in any case for assigning.

**Comment by ppedrot** [comment]
@herbelin you can still review and validate this PR though.

**Review by herbelin — APPROVED** [review]
> @herbelin you can still review and validate this PR though.

Sure. And let me seize this opportunity to encourage the Coq developers to not be satisfied with the current state of tactics and to consider seriously spending time at improving tactics (and I hope @ybertot and @mattam82 will agree). Let me highlight one typical example visible in the PR. The change from:
```
In a (x::l) := (x = a) \/ In x l
```
to
```
In a (x::l) := (a = x) \/ In x l
```
implies that some `apply H`, with `H` of type an equality, have to be replaced by `symmetry; apply H`. In my opinion, this is not moral. I see no reason for an `apply H` with `H` an equality to be sensitive to the order in which the equality is presented. From a human point of view, this is an irrelevant low-level worry of the `trivial` kind, and an obvious cause of non-robustness in scripts.

The PR is otherwise ok. If @ybertot has time, that would help to have a temporary compatibility `Notation` for `Rlength` I guess.

**Comment by ybertot** [comment]
Can you point me to an example of another PR where a temporary deprecated notation is added?

**Comment by ppedrot** [comment]
I am not even sure this deserves a compat notation, to be fair. This is a niche library that is virtually unused. But in any case it cannot hurt.

File: theories/Reals/RList.v
Comment (herbelin) [inline theories/Reals/RList.v]: @ybertot: I did not test but I believe that this should work:
```suggestion

#[deprecated(since="8.12",note="Use List.length instead.")] Notation Rlength := length.

```

@ppedrot:

> This is a niche library that is virtually unused

I'm not able to tell. It was used at least in Coquelicot...

One point which pushed me to propose this alias was the presence of the alias `Notation Rlist := (list R).`. Maybe `Rlist` should be abandoned too. After all, it saves only one character relatively to `list R` with `list R` immediately understandable while, when seeing `Rlist`, one may wonder what is so specific to the definition of `Rlist` which makes that `list R` has not been used. Let me propose another suggestion...

File: theories/Reals/RList.v
Comment (herbelin) [inline theories/Reals/RList.v]:
```suggestion
#[deprecated(since="8.12",note="Use (list R) instead.")] Notation Rlist := (list R).
```

**Comment by ppedrot** [comment]
@ybertot IIUC this is essentially ready. Can you squash the commits though? I don't think it brings much to keep them separate.

**Comment by ybertot** [comment]
will do

**Comment by ybertot** [comment]
I just squashed and rebased.  I am waiting for ci to succeed to send a merge request for coquelicot's overlay.

**Comment by ybertot** [comment]
After discussion with @silene concerning the coquelicot overlay, I wish to explore another possibility that may avoid the need for one.  Please consider this PR as WIP until I come back with the results.

**Comment by herbelin** [comment]
> I just squashed and rebased.

Great.

> Please consider this PR as WIP until I come back with the results.

No problem, ok.

**Comment by ybertot** [comment]
I tried adding the following line:
`#[deprecated(since="8.12",note="use List.cons instead")]
Notation cons := List.cons.`
with the hope that it would make coquelicot compile without the need for an overlay.  If the notation `RList.cons` was available in coquelicot, we suspect that the master branch of coquelicot (as of today) could be compatible with this PR.

It fails.  I believe it is because `cons` is already a notation defined in `List`.  Is there a way to have coquelicot compile both with coq-8.11 and this PR?

**Comment by SkySkimmer** [comment]
How does it fail? Shouldn't it be `Notation cons := (List.cons R).`?

**Comment by ybertot** [comment]
I just retried and I fear I had made a copy/paste mistake.  With the following four lines, it does not fail.

    #[deprecated(since="8.12",note="use List.cons instead")]
    Notation cons := List.cons.

    #[deprecated(since="8.12",note="use List.nil instead")]
    Notation nil := List.nil.

(I think my mistake is that I tried to define a `cons` notation twice.)

**Comment by ppedrot** [comment]
@ybertot the compatibility notations are triggering warnings in the compilation of `Rlist`, making the CI fails.

**Comment by ybertot** [comment]
I just pushed a new version that avoids the deprecated notations we just introduced.  When the deprecated notation is eventually removed, all instances of `List.nil` should be replaced by `nil` in the text of `RList.v`.  Do you thing this should be expressed in a comment?

**Comment by herbelin** [comment]
>  When the deprecated notation is eventually removed, all instances of List.nil should be replaced by nil in the text of RList.v. 

I'm unsure it is needed to use `List.nil` in `Rlist.v` rather than `nil` (which a priori already binds to `List.nil` via the notation).

> Do you thing this should be expressed in a comment?

If you prefer to use `List.nil`, then probably  a comment on what to do at removal of the notation would help. If using `nil` now works, no need.

File: theories/Reals/RList.v
Comment (SkySkimmer) [inline theories/Reals/RList.v]: If you put the compat notations at the end of the file you won't need to qualify `nil`/`cons` to avoid the warning.

**Comment by ybertot** [comment]
I just squashed and rebased again.  The new version should be compatible with the master branch of coquelicot: the overlay I had designed is not necessary and will be integrated in coquelicot later (when compatibility with coq-8.11 can be dropped)

**Comment by ppedrot** [comment]
This looks good to me. Everybody fine with the current state of the PR? I'll merge soon.

**Pull request merged**
